### PR TITLE
Add file/episode duration and resume position

### PR DESCRIPTION
**Pull request by revam**
Because we want to render the percentage some places in Web UI.

Example from @ElementalCrisis's mock-ups;
image

### CODE REVIEW

**Comment by revam** [comment]
Will merge late Sunday (local time) unless somebody wants something changed.

**Review by Cazzar — CHANGES_REQUESTED** [review]

File: Shoko.Server/API/v3/Controllers/DashboardController.cs
Comment (revam) [reply Shoko.Server/API/v3/Controllers/DashboardController.cs:301]: It was deliberate.

Comment (revam) [reply Shoko.Server/API/v3/Controllers/DashboardController.cs:301]: Though yes, it's unrelated and should not have been included in this PR probably.

Comment (Cazzar) [reply Shoko.Server/API/v3/Controllers/DashboardController.cs:301]: My point still stands though :)

File: Shoko.Server/API/v3/Models/Shoko/Dashboard.cs
Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/Dashboard.cs:132]: I added the the new constructor with the parameters in rising order from the smallest entry to the largest; so the file user data, then the file, then the episode, then the anime, and last the shoko series.

Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/Dashboard.cs:132]: It can be changed if you really want to.

File: Shoko.Server/API/v3/Models/Shoko/Dashboard.cs
Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/Dashboard.cs:140]: We can include the duration and resume position instead and leave it to the front end. Actually, that may be the better idea.

File: Shoko.Server/API/v3/Models/Shoko/Dashboard.cs
Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/Dashboard.cs:195]: Will leave it be then.

Forgot to prefix it when it was added and afaik it's not used yet since it was introduced recently.

File: Shoko.Server/API/v3/Models/Shoko/Episode.cs
Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/Episode.cs:58]: Will change it.

File: Shoko.Server/API/v3/Controllers/DashboardController.cs
Comment (revam) [reply Shoko.Server/API/v3/Controllers/DashboardController.cs:267]: Will change it.

**Comment by revam** [comment]
Tried to touch as little as possible in the new files while adding the server model (and removing any references to the non-server model) for the user records.

File: Shoko.Server/API/v3/Models/Shoko/Dashboard.cs
Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/Dashboard.cs]: Changed.

**Comment by revam** [comment]
I did have one breaking change in the last rebase; the `ResumePosition` for the `File` was changed from a `long` to a `TimeSpan`. But afaik then no current or previous client have used the field to date.

**Comment by bigretromike** [comment]
are there any v3 endpoints that got broken or old code should work without issue ?

**Comment by revam** [comment]
@bigretromike see my last comment/message. That's the only breaking change.

**Comment by revam** [comment]
Is Nakamori still using api v2?

**Comment by bigretromike** [comment]
> > Is Nakamori still using api v2?
> 
> It's not used in the v3 data models used in Nakamori it seems, so I still don't think the field have ever been used before.

we did v3 some time ago. If those changes break things we will need to rewrite some code, thats why im asking.
Then again, why not make another v3 version like it was done before. like v3.1 its a good thing to not break compatible.

**Comment by revam** [comment]
And i'm saying i know what I broke (and when it was introduced) and I am not familiar with any current or previous clients using the field within the last ~2 years since it was introduced. Also i will rather revert the type to a `long` than create another api sub version.

**Comment by ElementalCrisis** [comment]
Even though APIv3 has been in the works for a while, it's still in dev and subject to breaking changes. As Nakamori is currently using it, maybe @revam you can tag @bigretromike anytime APIv3 is updated with a breaking change so he stays informed?

File: Shoko.Server/API/v3/Models/Shoko/Dashboard.cs
Comment (Cazzar) [reply Shoko.Server/API/v3/Models/Shoko/Dashboard.cs]: I would add new, optional parameters at the end instead of doing the override constructor.

**Comment by bigretromike** [comment]
> Even though APIv3 has been in the works for a while, it's still in dev and subject to breaking changes. As Nakamori is currently using it, maybe @revam you can tag @bigretromike anytime APIv3 is updated with a breaking change so he stays informed?

That would be awesome - that or just add changelog with information that function is not backward compatible anymore.

@ElementalCrisis it's been in work as was v2 before, Shoko API won't ever be `finished` its always something that will evolve with rest of the code. You guys introduce sub api versions, and imo thats best place to put anything that otherwise break compatible - but im not forcing that anyway.

**Comment by da3dsoul** [comment]
Sure things are constantly changing, but we've done a really good job of not breaking apiv2

**Comment by Cazzar** [comment]
Api sub versions are nice but for a small change like this it is entirely overkill and not worth adding it.
When the discussion was made about the potentially breaking change, we decided to accept that risk as APIv3 is as @ElementalCrisis mentioned, still in its “active development/unfinished” state unlike apiv1/apiv2

File: Shoko.Server/API/v3/Models/Shoko/Dashboard.cs
Comment (revam) [reply Shoko.Server/API/v3/Models/Shoko/Dashboard.cs]: Removed the secondary constructor and added the parameters in the order of most importance (so the least important ones can be omitted).

**Comment by da3dsoul** [comment]
I think it's fine

**Pull request merged**
